iroh-net/src/hp/netcheck/reportgen/hairpin.rs

```diff
@@ -0,0 +1,318 @@
+//! Actor to run hairpinning check.
+//!
+//! This actor works as follows:
+//!
+//! - After starting prepares the haircheck:
+//!   - binds socket
+//!   - sends traffic from it's socket to trick some routers
+//! - When requested performs the hairpin probe.
+//!   - result is sent to netcheck actor addr.
+//! - Shuts down
+//!
+//! Note it will only perform a single hairpin check before shutting down.  Any further
+//! requests to it will fail which is intentional.
+
+use std::net::SocketAddr;
+use std::ops::Deref;
+use std::sync::Arc;
+use std::time::Duration;
+
+use anyhow::{bail, Context, Result};
+use tokio::net::UdpSocket;
+use tokio::sync::{mpsc, oneshot};
+use tokio::time::Instant;
+use tracing::{debug, error, instrument, trace, warn};
+
+use crate::hp::netcheck::{self, reportgen, Inflight};
+use crate::hp::stun;
+use crate::util::CancelOnDrop;
+
+/// The amount of time we wait for a hairpinned packet to come back.
+const HAIRPIN_CHECK_TIMEOUT: Duration = Duration::from_millis(100);
+
+/// Handle to the hairpin actor.
+///
+/// Dropping it will abort the actor.
+#[derive(Debug, Clone)]
+pub(super) struct Client {
+    addr: Addr,
+    _drop_guard: Arc<CancelOnDrop>,
+}
+
+impl Client {
+    pub(super) fn new(netcheck: netcheck::Addr, reportstate: reportgen::Addr) -> Self {
+        let (msg_tx, msg_rx) = mpsc::channel(32);
+        let mut actor = Actor {
+            msg_tx,
+            msg_rx,
+            netcheck,
+            reportstate,
+        };
+        let addr = actor.addr();
+        let task = tokio::spawn(async move { actor.run().await });
+        Self {
+            addr,
+            _drop_guard: Arc::new(CancelOnDrop::new("hairpin actor", task.abort_handle())),
+        }
+    }
+
+    /// Starts the hairpin check.
+    ///
+    /// *dst* should be our own address as discovered by STUN.  Hairpin detection works by
+    /// sending a new STUN request to our own public address, if we receive this request
+    /// back then hairpinning works, otherwise it does not.
+    ///
+    /// Will do nothing if this actor is already finished or a check has already started.
+    pub(super) fn start_check(&self, dst: SocketAddr) {
+        self.addr.try_send(Message::StartCheck(dst)).ok();
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Addr {
+    sender: mpsc::Sender<Message>,
+}
+
+impl Deref for Addr {
+    type Target = mpsc::Sender<Message>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.sender
+    }
+}
```

Comment:
True, though it only removes some lines of code in return for having to look up a bit more how macros work.  I like derive_more's Debug macro since it solves a problem: new fields on a struct don't get forgotten to be added to the manual Debug impl.  But this is just another way or writing things so I'm rather non-plussed about it (this ties in with my general dislike for macros :smile:, it's another language you have to learn)